scipy/spatial/transform/_rotation.pyx

```diff
@@ -223,6 +236,140 @@ cdef double[:, :] _compute_euler_from_matrix(
 
     return angles
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef double[:, :] _compute_euler_from_quat(
+    np.ndarray[double, ndim=2] quat, const uchar[:] seq, bint extrinsic=False
+):
+    # The algorithm assumes extrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation quaternions, which are stored
+    # directly by Rotation.
+    # Adapt the algorithm for our case by
+    # 1. Replacing "d" with "-d" for intrinsic rotation
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+    
+    if not extrinsic:
+        seq = seq[::-1]
+
+    cdef int i = _elementary_basis_index(seq[0])
+    cdef int j = _elementary_basis_index(seq[1])
+    cdef int k = _elementary_basis_index(seq[2])
+    
+    # quick renormalization of transformed quaternion, assumes |q| = 1
+    cdef int norm_squared 
+
+    is_proper = i == k
+    if is_proper:
+        k = 3 - i - j # get third axis
+        norm_squared = 1
+    else:
+        norm_squared = 2
+        
+    # Step 0
+    # Check if permutation is even (+1) or odd (-1)     
+    cdef int sign = (i-j)*(j-k)*(k-i)//2
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double a, b, c, d, cos_theta_2
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef bint safe1, safe2, safe, adjust
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+
+        # Step 1
+        # Permutate quaternion elements            
+        if is_proper:
+            a = quat[ind, 3]
+            b = quat[ind, i]
+            c = quat[ind, j]
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+        
+        # Step 2
+        # Ensure less than unit norm
+        cos_theta_2 = 2*(a**2 + b**2)/norm_squared - 1
+        cos_theta_2 = min(cos_theta_2, 1)
+        cos_theta_2 = max(cos_theta_2, -1)
+        
+        # Step 3
+        # Compute second angle...
+        _angles[1] = acos(cos_theta_2)
```

Comment:
A question more related to math. 

As far as I understand `acos` near 0 and 1 suffer somewhat for numerical issues (due to expansion `cos(x) ~ 1 - x**2 / 2`). Like `np.arccos(np.cos(1e-16)) == 0.0`.

Is it possible to replace `acos` with `atan2` here? Judging by the matrix algorithm - probably not, but worth considering.

Reply:
Equations (34) and (35) of the article actually give 4 different formulas for calculating this! They are:

1. _angles[1] = 2*acos( sqrt( (a**2 + b**2) / norm_squared ) )`
2. _angles[1] = 2*asin( sqrt( (c**2 + d**2) / norm_squared ) )`
3. _angles[1] = 2*atan( sqrt( (c**2 + d**2)/(a**2 + b**2) ) )`
4. _angles[1] = acos( 2 * (a**2 + b**2) / norm_squared - 1 )`

The third equation has the advantage of not needing the calculation of `norm_squared`. In theory, we are assuming that `_angles[1]` is between 0 and pi, so instead of using `atan2`, I just used `atan` since both arguments are positive and should be equivalent.

What made me choose the fourth equation, though, is that using this one, there is no need to calculate any **square roots**, which I imagine might be desirable.

If you prefer, though, I can replace this code for the `atan` formula and remove the need of `norm_squared` in the meantime.